scipy/stats/tests/test_stats.py

```diff
@@ -6615,6 +6615,24 @@ def test_mudholkar_george_equal_fisher_minus_pearson(self):
         # 0.5 here is because logistic = log(u) - log(1-u), i.e. no 2 factors
         assert_approx_equal(0.5 * (Z_f-Z_p), Z, significant=4)
 
+    @pytest.mark.parametrize("variant", ["single", "all"])
+    @pytest.mark.parametrize("method",
+                             ["fisher", "pearson", "tippett",
+                              "stouffer", "mudholkar_george"])
+    def test_monotony(self, variant, method):
+        # Test that result increases monotonously with respect to input.
+        changing_values = np.linspace(0.1, 0.9, 10)
+        rng = np.random.default_rng(278448169958891062669391462690811630763)
```

Comment:
Oh, sorry I missed that. I had received a conflicting suggestion before and was passing it on. I didn't realize that the recommended way had changed.

Reply:
For testing purposes, there is nothing wrong per se with 1234. It's yet again a way to prevent users to copy paste seeds in their codes. Note that from the RNG perspective, small seeds are bad for initialization. This is why NumPy recommends using `np.random.SeedSequence().entropy` when there is a need for a seed.

Reply:
I see, thanks for the explanation, it makes sense. I've never actually looked into the details of how NumPy's random number generators work. I think some will run a fixed number of iterations from the initial seed before using the output to get around such issues, although maybe that's still not good enough. For actual research practice I have one concern though: how are we to know that researchers have actually used `np.random.SeedSequence().entropy` properly rather than for example using it repeatedly until say their proposed machine learning algorithm achieves state of the art results on a set of benchmarks? Obviously that's not a concern for this PR though.

Reply:
> how are we to know that researchers have actually used np.random.SeedSequence().entropy properly rather than for example using it repeatedly until say their proposed machine learning algorithm achieves state of the art results on a set of benchmarks? 

That's exactly what I don't like about being able to seed 😉 I've seen a lot of academic or production code with generators fixed to a given seed. Then you cannot say anymore that you are using a random number generator, it's just a fixed sequence which seems to be random.

Personally, I prefer to differentiate random generators and sequences. For cases where you need to have some reproducibility of the stream of number, I would use low discrepancy sequence (because the stream is of better quality than a MC with a fixed seed and also not platform/implementation dependent, etc.) and when you need random numbers then forbid seeding of MC or use scrambled sequences. This way it's clearer and you're less prone to error/bias.

But to your question, in the end, people will always find a way to manipulate the results to their advantage... 😅

Reply:
> hat's exactly what I don't like about being able to seed wink I've seen a lot of academic or production code with generators fixed to a given seed. Then you cannot say anymore that you are using a random number generator, it's just a fixed sequence which seems to be random.

Interesting. There's one case where I've used an equidistributed sequence in a quasi-monte-carlo algorithm, but I'd never considered them as a replacement for psuedorandom numbers before.  For randomized algorithms, I think what one actually desires is a fixed sequence which seems to be random. Von Neumann had this to say on the matter:

*The real objection to this procedure is the practical need for checking computations. If we suspect that a calculation is wrong, almost any reasonable check involves repeating something done before. At that point the generation of new random numbers would be intolerable.* [1]

There's also cases where using a fixed sequence reduces unwanted variability. Such as when selecting cross-validation splits for a data-set. Reproducibility is really important. I guess I need to look into the literature more on how different randomized algorithms are affected when using a low discrepancy sequence. I'd be really cautious about the idea of ever using non-reproducible sequences of random numbers though.



[1] Von Neumann, John (1951). ["Various techniques used in connection with random digits"](https://dornsifecms.usc.edu/assets/sites/520/docs/VonNeumann-ams12p36-38.pdf) (PDF). National Bureau of Standards Applied Mathematics Series. 12: 36–38.

Reply:
Ah yes, I remember this article. Thanks for reminding me of it, it's a great read!

> There's also cases where using a fixed sequence reduces unwanted variability. Such as when selecting cross-validation splits for a data-set. Reproducibility is really important.

I think that this is a good use case for low discrepancy sequences. They provide reproducible sequences of numbers but properly sample the parameter space (this is even worse for MC if you would sample in n-D), which is what you want when splitting data-set. With MC you have a lot of variance in the quality of the sample and could pick a very good seed or a very bad one from one seed to another. Using sequences also helps a lot if you really care about reproducibility since the stream of numbers is defined and known.

Here I am comparing QMC sequences to things like `np.random.default_rng(0)` or `np.random.default_rng(1234)` which are widely used.

Reply:
> I think that this is a good use case for low discrepancy sequences.

Yeah, I already agree there actually. Von Neumann goes on to say in that paper that being able to check calculations in case of mistakes is the **only** reason to prefer reproducible sequences, so I just wanted to cover my bases with a class of reasons he didn’t foresee.

Reply:
I’m really curious how well something like a random forest would perform with psuedorandom numbers vs a low-discrepancy sequence. My intuition says that pseudorandom would work better, but I want to check over the weekend.

Reply:
I am interested by things like this, if you could keep me updated 😃 I am part of the organization of a QMC conference (MCQMC2022) and this is something that could be of interest.